src/compiler/checker.ts

```diff
@@ -5345,72 +5333,61 @@ namespace ts {
+         * For example, property members with types inferred from initializers or function members with inferred return types are
+         * conservatively assumed to reference `this`.
          */
-        function isThislessFunctionLikeDeclaration(node: FunctionLikeDeclaration): boolean {
-            const returnType = getEffectiveReturnTypeNode(node);
-            return (node.kind === SyntaxKind.Constructor || (returnType && isThislessType(returnType))) &&
-                node.parameters.every(isThislessVariableLikeDeclaration) &&
-                (!node.typeParameters || node.typeParameters.every(isThislessTypeParameter));
-        }
-
-        /**
-         * Returns true if the class or interface member given by the symbol is free of "this" references. The
-         * function may return false for symbols that are actually free of "this" references because it is not
-         * feasible to perform a complete analysis in all cases. In particular, property members with types
-         * inferred from their initializers and function members with inferred return types are conservatively
-         * assumed not to be free of "this" references.
-         */
-        function isThisless(symbol: Symbol): boolean {
-            if (symbol.declarations && symbol.declarations.length === 1) {
-                const declaration = symbol.declarations[0];
-                if (declaration) {
-                    switch (declaration.kind) {
-                        case SyntaxKind.PropertyDeclaration:
-                        case SyntaxKind.PropertySignature:
-                            return isThislessVariableLikeDeclaration(<VariableLikeDeclaration>declaration);
-                        case SyntaxKind.MethodDeclaration:
-                        case SyntaxKind.MethodSignature:
-                        case SyntaxKind.Constructor:
-                            return isThislessFunctionLikeDeclaration(<FunctionLikeDeclaration>declaration);
-                    }
+        function symbolMayReferenceThis(symbol: Symbol): boolean {
+            const declaration = singleOrUndefined(symbol.declarations);
+            if (!declaration) return true;
+            switch (declaration.kind) {
+                case SyntaxKind.PropertyDeclaration:
+                case SyntaxKind.PropertySignature:
+                    return variableLikeDeclarationMayReferenceThis(<PropertyDeclaration | PropertySignature>declaration);
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.MethodSignature:
+                case SyntaxKind.Constructor: {
+                    // A function-like declaration references `this` if its return type does or some parameter / type parameter does.
+                    const fn = declaration as MethodDeclaration | MethodSignature | ConstructorDeclaration;
+                    return (declaration.kind !== SyntaxKind.Constructor && typeMayReferenceThis(getEffectiveReturnTypeNode(fn)))

```

Comment:
if `getEffectiveReturnTypeNode` returns `undefined` for nodes with `SyntaxKind.Constructor`, then you can skip the kind check here.

Reply:
#20044